src/components/Modal/Modal.tsx

```diff
@@ -0,0 +1,159 @@
+import * as React from 'react';
+import * as ReactDOM from 'react-dom';
+import createFocusTrap from 'focus-trap';
+import { st, classes } from './Modal.st.css';
+import { TPAComponentsConsumer } from '../TPAComponentsConfig';
+import { ModalProps, ModalDefaultProps, ModalState } from './types';
+import { CloseIcon } from './Close';
+
+export class Modal extends React.Component<ModalProps, ModalState> {
+  static displayName = 'Modal';
+  static defaultProps: ModalDefaultProps = {
+    isOpen: false,
+    closeOnClickOutside: true,
+    minHeight: '150px',
+    maxHeight: '620px',
+    minWidth: 'auto',
+    maxWidth: '580px',
+    focusTrap: true,
+  };
+
+  el = null;
+  focusTrapInstance = null;
+
+  constructor(props: ModalProps) {
+    super(props);
+
+    this.el = document.createElement('div');
+    this.state = {
+      isCloseInProgress: false,
+    };
+  }
+
+  componentDidMount() {
+    this.props.rootElement.appendChild(this.el);
+    if (this.props.focusTrap) {
+      this.focusTrapInstance = createFocusTrap(this.el, {
+        escapeDeactivates: false,
+        clickOutsideDeactivates: false,
+        returnFocusOnDeactivate: false,
+      });
+    }
+  }
+
+  componentDidUpdate(prevProps: ModalProps) {
+    if (this.props.isOpen !== prevProps.isOpen && this.focusTrapInstance) {
+      if (this.props.isOpen) {
+        try {
+          this.focusTrapInstance.activate();
+        } catch (err) {
+          console.error(err);
+        }
+      } else {
+        this.focusTrapInstance.deactivate();
+      }
+    }
+  }
+
+  onClose = () => {
+    this.setState(
+      {
+        isCloseInProgress: true,
+      },
+      () => {
+        setTimeout(() => {
+          this.setState(
+            {
+              isCloseInProgress: false,
+            },
+            () => {
+              if (this.props.onClose) {
+                this.props.onClose();
+              }
+            },
+          );
+        }, 200);
+      },
+    );
```

Comment:
Don't forget that if we use `setTimeout` we need to be able to `clearTimeout` in order to avoid exceptions and memory leaks.
Also, I think instead of putting all this logic in the close method, I would do something like this:
```tsx
_onClose = () => {
  this.setState({ isCloseInProgress: true });
}
```

and add to `componentDidUpdate`:
```tsx
componentDidUpdate(prevProps: ModalProps) {
  ...
  if (!prevProps.isCloseInProgress && this.props.isCloseInProgress) {
    clearTimeout(this._closingTimeout);
    this._closingTimeout = setTimeout(() => {
      this.setState({isCloseInProgress: false});  
    }, 200);
  } else if (prevProps.isCloseInProgress && !this.props.isCloseInProgress) {
    if (this.props.onClose) {
      this.props.onClose();
    }
  }
}
```

and also always we add a `clearTimeout` to `componentWillUnMount`